### ISSUE TITLE
Epic: layer eviction

### ISSUE THREAD

**Issue by stepashka** [neondatabase/neon#2476]
## DoD

```[tasklist]
### Tasks
- [ ] #3516
- [ ] #3540
- [ ] #3574
- [ ] #3552
- [ ] #3570
- [ ] #3557
- [ ] #3496
- [ ] #3504
- [ ] #3544
- [ ] #3527
- [ ] #3575
- [ ] #3533
- [x] #3439
- [ ] https://github.com/neondatabase/neon/issues/3585
```

LOWER PRIORITY
Figure out how to do tenant size computation without downloading remote layers (logical more tricky than physical, we don't have the metadata in index_part.json).
If we don't do this, currnent logical size on-demand download will redownload layers, defeating the point of layer file eviction to S3.

## Other related tasks and Epics
-

**Comment by vadim2404** [comment]
@hlinnaka , provide a description for this task

**Comment by hlinnaka** [comment]
The download part is tracked by https://github.com/neondatabase/neon/issues/2029. The part that remains after that is the layer eviction.

In short, layer eviction means deleting layer files from local disk in the pageserver, to free up disk space. All the files on local disk are also present in the remote storage, and with the on-dmenad download, if a file is needed again, it will be automatically downloaded. This allows us to handle larger databases, and in particular databases with lots of "history", that would not otherwise fit in a single pageserver. It also allows us to pack more tenants in a pageserver, saving resources.

**Comment by hlinnaka** [comment]
Some notes on implementing layer eviction:

## Eviction policy

Lots of options. Could be a simple Least-Recently Used policy. Or something that prioritizes more recent layer files, or something that tries to ensure some fairness between tenants, preferring to keep the most important files for each tenant present.

For now, I'm thinking that a very simple LRU policy, global across all tenants, is good enough. If we're not happy with that, we can improve on it later, once we have more experience with how it behaves and what problems we encounter.

The cache policy is more important, the less disk we have available in the pageserver. On the other extreme, if the pageserver has enough disk space to hold all files on local disk, nothing needs to be evicted. That's the situation we have as of this writing anyway, as we don't have on-demand download yet. We can go slowly, having lots of local disk available in the pageservers at first, but get more aggressive with the eviction and disk allocation over time.

The pageserver can utilize all the disk space that is available, there is no other use for the disk space currently. So the eviction can be triggered by disk usage alone. Let's add a configuration option for max local disk usage, set that to ~90 % of the disk size on the server to leave some headroom, and trigger eviction when that size is reached.

## Implementation notes

We must be careful not to evict a layer that hasn't been uploaded to remote storage yet. We currently have that information in the upload queue of the RemoteTimelineClient, but that's not very convenient for the eviction purposes. I think we need to add a new state field to Layer struct, to indicate if the Layer file is present in remote storage or not.

Currently, the layer map that tracks local layer files is a per-timeline struct. To implement a global LRU list, we'll need some global list of Layers too. Perhaps it can be simply one global `LinkedList<Arc<dyn Layer>>` or `LinkedList<Weak<dyn Layer>>`.

**Comment by shanyp** [comment]
We need:
statistics collection - per activity  (**access count / Activity / last access**) , collected by API. This will be used to analyze if LRU will work 

Global layer map data

Layer eviction API and statistics API

LRU implementation 

Metrics for LRU (cache miss / cache hit)

**Comment by problame** [comment]
My notes from yesterday's meeting:

**Goal**: the system should evict layers that haven't been accessed for a week.

**Non-goal**: the system regularly swaps layers in and out of remote storage to accommodate a maximum resident-physical-size setting.

The consequence:
We will not evict layers on demand when disk space is low and we need more. Instead, we will continue to run out of disk space. This behavior is preferable because it prevents thrashing.

Milestones:

1. Facility to monitor layer accesses.
   - most recent access time
   - the reason for access (page_service, gc, compaction, size calculation, ...)
2. Roll out monitoring facility to staging & production.
3. Use the monitoring facility in production and validate that there actually _are_  layer files that have not been accessed for a week.
  3.1. If there are no such layer files, we need to change the pageserver code to stop making those accesses.
       Reproduce the behavior on staging, then iterate by changing code and using the monitoring facility to confirm the effectiveness of the changes.
4. Add a facility to evict layer files manually
5. Do manual eviction experiments. Example:
   - Pick a busy tenant
   - Use monitoring data to identify victim layers (e.g., by looking at the most recent access time)
   - Evict those layers using the manual eviction facility
   - Watch logs &  performance Prometheus metrics
6. Automatic eviction:
   - the access stats added by the monitoring facility should suffice for decision making
   - policy need not be perfect LRU; we can iterate over all layers in the system and evict based on a configured threshold, e.g., "layer has not been accessed for X days"

Development Roadmap:


1. In parallel:
   - Build the monitoring facility
     - Merge `RequestContexts` to get "reason for access" https://github.com/neondatabase/neon/pull/3228
     - Implement monitoring facility code.
       - Maybe feature-flag metrics collection code if it has perf overhead.
     - Including tools to scrape the metrics & post-process them if necessary
      - Think of tools to visualize the data over time, produce a list of eviction candidates based on last access time, etc.
   - Build the manual eviction mechanism
1. Code changes required for Milestone 3.1
1. Code changes required in response to manual eviction experiments (we're likely going to hit some bugs that we want to address before automating eviction)
1. Automatic eviction implementation

**Comment by problame** [comment]
My notes from today's meeting:

- Concern: eviction means the pageserver disk space is overcommitted => risk of sudden usage spike => disk full
  - v1 doesn't aim to solve this problem perfectly, as the best we can do at the pageserver level is to start thrashing, which is not desirable.
  - for the first release, on-call will have to move tenants to different pageservers, or create new ones if necessary
  - mid-term, pageservers or control plane will need to automate this

- Observation: Dmitry's migration of the last pageservers into the new AWS account will use on-demand download.
  Unlike the previous migrations, which happened before on-demand download, we can plot resident vs. remote size.
  After the tenants are used for a few days, and there is still a gap between resident & remote even after tenant usage, that will be an approximation
  for the potential resident storage savings by layer eviction.
  - The monitoring facility will still be necessary, though, so the development roadmap stays unchanged in that regard.

- The manual eviction mechanism must come with two endpoints: one to dump the current layer map and one to trigger the eviction of a layer.
  We can use these endpoints for
  - functional testing of eviction mechanism in unit tests/ staging/prod
  - prototyping eviction policies as a sidecar service on the pageservers
    - Example: for our benchmark projects, we can have a python script that runs just for that project's timeline. It would watch the layer map through the endpoint and evict layers according to the eviction policy we want to evaluate. We'll run that script as a service on the pageserver, then watch the benchmark's performance over time.
    - This allows for significantly faster iteration speed than baking one policy into the pageserver & feature-gating it.

- Regarding automatic eviction (roadmap item 4): we'll still need LRU, but layers older than X days should never be considered victims.
  - We probably want X to be configurable per timeline to be able to react to perf problems of some tenants quickly(?)

**Comment by shanyp** [comment]
removed it and closed
